core/modules/wcontrol/RunnerMgr.h

```diff
@@ -0,0 +1,72 @@
+// -*- LSST-C++ -*-
+
+#ifndef LSST_QSERV_WCONTROL_RUNNERMGR_H
+#define LSST_QSERV_WCONTROL_RUNNERMGR_H
+
+// System headers
+#include <atomic>
+#include <memory>
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// Local headers
+#include "wcontrol/Foreman.h"
+
+namespace lsst {
+namespace qserv {
+namespace wdb {
+	class QueryRunner;
+}}}
+
+namespace lsst {
+namespace qserv {
+namespace wcontrol {
+
+/// Run an event driven thread that is fed by the Foreman's Scheduler.
+class Runner : public std::enable_shared_from_this<Runner> {
+public:
+	using Ptr = std::shared_ptr<Runner>;
+    Runner(RunnerMgr& rm, wbase::Task::Ptr firstTask);
+    Runner(Runner&) = delete;
+    Runner operator=(Runner&) = delete;
+    virtual ~Runner();
+    void operator()();
+    std::string const& getHash() const { return _task->hash; }
+
+private:
+    RunnerMgr& _rm;
+    wbase::Task::Ptr _task;
+    std::atomic<bool> _poisoned{false}; // TODO:remove in DM-3945
+};
+
+/// Keep track of the existing Runner objects for the Foreman.
+///
+class RunnerMgr {
+public:
+    RunnerMgr(Foreman& f);
+    RunnerMgr(RunnerMgr&) = delete;
+    RunnerMgr operator=(RunnerMgr&) = delete;
+    void registerRunner(Runner::Ptr const& r, wbase::Task::Ptr t);
```

Comment:
There were actually a few issues with Runner and RunnerMgr that all stemmed from 

    std::thread thrd{Runner{*_rManager, t}};

Runner had to be copyable for the above to work, which caused problems. The following allows Runner to be non-copyable and allows its costructor to take `Task::Ptr const& t` as an argument, but if you use `wbase::Task::Ptr const& t` as an argument in the lambda function, you can't use const& in the Runner constructor.

        auto f = [this](wbase::Task::Ptr t) {
            std::unique_ptr<Runner> rp{new Runner{*(this->_rManager), t}};
            (*rp)();
        };
        std::thread thrd{f, t};
        thrd.detach();